### PR TITLE
Create directory

### PR DESCRIPTION
**Pull request by sheepy0125**
Addresses #768.
Directories can be created with the `/upload` endpoint, with a `mkdir` multipart form key.
Attached is a video of it in use.

https://user-images.githubusercontent.com/73724119/164916100-4c537b0f-d305-4686-887d-5c2805f96e63.mp4

### CODE REVIEW

**Comment by svenstaro** [comment]
From a quick look, this look like great work! I'll take a more thorough look later.

**Comment by sheepy0125** [comment]
Sorry about the failed checks, hopefully it should work now...

File: tests/create_directories.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
+mod fixtures;
+
+use fixtures::{server, Error, TestServer};
+use reqwest::blocking::{multipart, Client};
+use rstest::rstest;
+use select::document::Document;
+use select::predicate::{Attr, Text};
+
+#[rstest]
+fn creating_directories_works(#[with(&["-u", "-U"])] server: TestServer) -> Result<(), Error> {
```
Comment (svenstaro) [inline tests/create_directories.rs:10]: Use the long flags here.

File: tests/create_directories.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    let parsed = Document::from_read(body)?;
+    assert!(parsed
+        .find(Text)
+        .any(|x| x.text() == test_directory_name.to_owned() + "/"));
+
+    Ok(())
+}
+
+#[rstest]
+fn creating_directories_is_prevented(server: TestServer) -> Result<(), Error> {
```
Comment (svenstaro) [inline tests/create_directories.rs:47]: Add a doc comment here describing _when_ or _why_ it is prevented.

File: tests/create_directories.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+/// see: https://github.com/svenstaro/miniserve/issues/518
+#[rstest]
+#[case("foo", "bar", "foo/bar")] // Not CWE-22, but `foo` isn't a directory
+#[case("/../foo", "bar", "foo/bar")]
+#[case("/foo", "/../bar", "foo/bar")]
+#[case("C:/foo", "C:/bar", if cfg!(windows) { "foo/bar" } else { "C:/foo/C:/bar" })]
+#[case(r"C:\foo", r"C:\bar", if cfg!(windows) { "foo/bar" } else { r"C:\foo/C:\bar" })]
+#[case(r"\foo", r"\..\bar", if cfg!(windows) { "foo/bar" } else { r"\foo/\..\bar" })]
+fn prevent_path_transversal_attacks(
+    #[with(&["-u", "-U"])] server: TestServer,
```
Comment (svenstaro) [inline tests/create_directories.rs:94]: Also use the long flags here.

**Review by svenstaro — CHANGES_REQUESTED** [review]
Great work! Just some small things and we're there. :)

**Comment by svenstaro** [comment]
Also make sure to resolve the conflicts/rebase. :)

**Comment by sheepy0125** [comment]
I have no clue how to rebase... I don't think I did it right

**Comment by svenstaro** [comment]
I'm afraid this doesn't look right :) You have two weird merges in there right now. Make sure your local `master` is up to date with upstream. Then, let's try an interactive rebase onto it while at the same time getting rid of the merges in there. I'd try to start something with `git rebase -i master~3`. There are plenty of great interactive tutorials on this.

**Comment by svenstaro** [comment]
While trying this out I noticed something: It's possible to specify `-U` without `-u` which will actually render a button but that button can't be used due to error 405. I think perhaps we should specify `-U` to require `-u`?

**Comment by sheepy0125** [comment]
Hmm... since it is using the `/upload` endpoint, it'd be easier to require `-U` to also have `-u`. I think that's a good idea. What should happen if `-U` is set without `-u`, though?

**Comment by svenstaro** [comment]
> What should happen if `-U` is set without `-u`, though?

Just let [clap deal with it](https://docs.rs/clap/latest/clap/struct.Arg.html#method.requires) and show the user an error.

File: src/args.rs [hunk 4]
```diff
@@ -110,9 +110,13 @@ pub struct CliArgs {
     /// Enable file uploading
     #[clap(short = 'u', long = "upload-files")]
     pub file_upload: bool,
+    
+    /// Enable creating directories
+    #[clap(short = 'U', long = "mkdir", requires("file-upload"))]
+    pub mkdir_enabled: bool,
 
     /// Specify uploadable media types
-    #[clap(arg_enum, short = 'm', long = "media-type", requires = "file-upload")]
```
Comment (svenstaro) [inline src/args.rs:115]: Hey, what's the reason for touching this one?

File: src/args.rs [hunk 5]
```diff
@@ -110,9 +110,13 @@ pub struct CliArgs {
     /// Enable file uploading
     #[clap(short = 'u', long = "upload-files")]
     pub file_upload: bool,
+    
+    /// Enable creating directories
+    #[clap(short = 'U', long = "mkdir", requires("file-upload"))]
```
Comment (svenstaro) [inline src/args.rs:115]: Can you use the `=` syntax?

**Review by svenstaro — CHANGES_REQUESTED** [review]

File: src/args.rs (hunk 4 above)
Comment (sheepy0125) [reply src/args.rs:115]: Sorry, I have no idea why I touched that.

**Comment by svenstaro** [comment]
Ok, one last issue I found with this is that you're currently apparently able to create directories in symlinks which point to directories even when symlinks are disabled. Coincidentally, I also found that it's currently possible to upload files to directories pointed to by symlinks when symlinks are disabled.

Would you like to take a stab at fixing that?

**Comment by sheepy0125** [comment]
That's not good... sure, I'll try!

**Comment by svenstaro** [comment]
Amazing job! Merging as is. Thanks for this great contribution. Since you're now very familiar with this code, especially the symlinking stuff, any chance you could take a look at #498?

**Pull request merged**